src/components/DataProvider/index.tsx

```diff
@@ -0,0 +1,199 @@
+import * as React from 'react';
+import { UpdateDomains } from '../Griff';
+import {
+  calculateDomains,
+  isEqual,
+  highestPriorityDomain,
+} from '../../utils/domains';
+import { deleteUndefinedFromObject } from '../../utils/cleaner';
+import { placeholder } from '../../utils/placeholder';
+import { LoaderResult, Domain, LoaderReason } from '../../external';
+import { ScaledSeries, ScaledCollection } from '../../internal';
+import { debounce } from 'lodash';
+import { Props as ScalerProps } from '../Scaler';
+import Joiner from '../Joiner';
+import getItemsByItemId from '../../utils/itemsById';
+
+export interface Props extends ScalerProps {
+  series: ScaledSeries[];
+  collections: ScaledCollection[];
+
+  updateDomains: UpdateDomains;
+}
+
+interface State {
+  loaderResultsById: {
+    [seriesId: string]: LoaderResult & {
+      dataDomains: { time: Domain; x: Domain; y: Domain };
+    };
+  };
+}
+
+type SeriesById = { [seriesId: string]: ScaledSeries };
+
+type SeriesByCollectionId = { [seriesId: string]: ScaledSeries };
+
+interface RequestRecord {
+  timestamp: number;
+  timeSubDomain: Domain;
+}
+
+type FetchFunction = (
+  series: ScaledSeries,
+  reason: LoaderReason
+) => Promise<void>;
+
+const PLACEHOLDER_SUBDOMAIN = placeholder(0, 0);
+
+const DEFAULT_DATA_DOMAINS = {
+  time: PLACEHOLDER_SUBDOMAIN,
+  x: PLACEHOLDER_SUBDOMAIN,
+  y: PLACEHOLDER_SUBDOMAIN,
+};
+
+class DataProvider extends React.Component<Props, State> {
+  inFlightRequestsById: { [seriesId: string]: RequestRecord } = {};
+
+  fetchFunctionsById: { [seriesId: string]: FetchFunction } = {};
+
+  state: State = {
+    loaderResultsById: {},
+  };
+
+  componentDidUpdate() {
+    const { series } = this.props;
+
+    series.forEach((s: ScaledSeries) => {
```

Comment:
Should we restrict this call only to series changes?

Reply:
I originally did that, but my thinking was that if a series didn't change, then it'll be a no-op to update it. Filtering them out could only lead to complications (and possibly a small performance boost).

I'm of the opinion that we can add this in later once/if there are performance problems -- what do you think?

Reply:
sure, feel free